src/App.tsx

```diff
@@ -15,49 +20,98 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+interface SortParams<T> {
+  sortReverse: boolean;
+  sortField: keyof T | '';
+}
+
+function getPreparedGoods<T>(
+  goods: T[],
+  { sortReverse, sortField }: SortParams<T>,
+) {
+  const preparedGoods = [...goods];
+
+  if (sortField) {
+    preparedGoods.sort((element1, element2) => {
+      const value1 = element1[sortField];
+      const value2 = element2[sortField];
+
+      if (typeof value1 === 'number' && typeof value2 === 'number') {
+        return value1 - value2;
+      }
+
+      if (typeof value1 === 'string' && typeof value2 === 'string') {
+        return value1.localeCompare(value2);
+      }
+
+      return 0;
+    });
+  }
+
+  if (sortReverse) {
+    return preparedGoods.reverse();
+  }
+
+  return preparedGoods;
+}
+
 export const App: React.FC = () => {
+  const [sortField, setSortField] = useState<keyof SortType | ''>('');
+  const [sortReverse, setSortReverse] = useState(false);
+  const visibleGoods
+  = getPreparedGoods(goodsFromServer, { sortReverse, sortField });
+
   return (
     <div className="section content">
       <div className="buttons">
         <button
           type="button"
-          className="button is-info is-light"
+          className={classnames(
+            'button is-info',
+            { 'is-light': sortField !== SortType.Name as keyof SortType },
+          )}
+          onClick={() => setSortField(SortType.Name as keyof SortType)}
         >
           Sort alphabetically
         </button>
 
         <button
           type="button"
-          className="button is-success is-light"
+          className={classnames(
+            'button is-success',
+            { 'is-light': sortField !== SortType.Length },
+          )}
+          onClick={() => setSortField(SortType.Length)}
         >
           Sort by length
         </button>
 
         <button
           type="button"
-          className="button is-warning is-light"
+          className={classnames(
+            'button is-warning',
+            { 'is-light': !sortReverse },
+          )}
+          onClick={() => setSortReverse(!sortReverse)}
```

Comment:
move to an own function and use a `prev` value 
```
const handleReverseClick = () => {
  setSortReverse(prev => !prev)
}
```
```suggestion
          onClick={handleReverseClick}
```